lib/ecto/changeset.ex

```diff
@@ -1161,8 +1161,10 @@ defmodule Ecto.Changeset do
   then falls back on the data, finally returning `default` if
   no value is available.
 
-  For relations, these functions will return the changeset data
-  with changes applied. To retrieve raw changesets, please use `get_change/3`.
+  For associations and embeds, this function returns the changeset
+  data with all changes applied. Use  `get_assoc/3`/`get_embed/3`
+  if you want to retrieve the relations as changesets or if you want
+  more fine-grained control.
```

Comment:
```suggestion
  data with all changes applied. If the association has not been loaded
  yet, we assume it is empty. Use `get_assoc/3`/`get_embed/3`
  if you want to retrieve the relations as changesets or if you want
  more fine-grained control.
```

Reply:
```suggestion
  If the changeset's data has not been loaded, we assume any association that has not been loaded
  yet is empty
```

It's slightly different. If the parent data is loaded it will raise if the association is not loaded. If the parent data is built it will assume a not loaded association is empty. And then the reason the latter might be an issue is because you can preload onto built structs as long as they have the primary key.

Reply:
`get_assoc` here only highlights what is already working the same with `get_field` or `put_assoc`. `get_assoc` by itself doesn't really "populate assocs", only a later `put_assoc` will. And `put_assoc` has always allowed you to put associations on a `:built` schema struct without an explicit preload needed. So to me it sounds like what you're describing here as "not ok" is actually the intended functionality. Both `get_field` and `put_assoc` don't care if associations are preloaded if the struct hasn't been persisted yet. Nothing about that changed with `get_assoc`.

Reply:
> %User{id: id} |> get_assoc(...) = not ok because will populate empty assoc

Ah, it was use to behave like this but it was a major pain because you had to do this:

```
%User{}
|> preload(...)
|> put_assoc(...)
```

for something you knew that was empty. And the only way to make those decisions reliably is based if the struct is loaded or not.

So I understand your point of consistency, but we would be enforce consistency for a weird behaviour (manually setting the primary key) at the cost of making it more annoying for everyone for a very common behaviour.

Another way to put it, if you are manually setting the `id` for a struct for updates, you are already starting from a point where the whole data in the struct does not match the database. Assocs/embeds are only part of the same issue. If you want to load directly, then I suggest to use Repo.load or similar.

Reply:
No docs changed needed, i agree.

Reply:
Oh I thought Jose already changed the docs last night based on the misunderstand we had. If he didn't then no change.